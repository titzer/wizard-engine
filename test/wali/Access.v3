import component wali {
	def SYS_open(pathname: Pointer, flags: i32, mode: i32) -> i64;
	def SYS_close(fd: i32) -> i64;
	def SYS_read(fd: i32, buf: Pointer, count: i32) -> i64;
	def SYS_write(fd: i32, buf: Pointer, count: i32) -> i64;
	def SYS_access(pathname: Pointer, mode: i32) -> i64;
}

def O_RDONLY = 0;
def STDOUT = 1;
def READ_SIZE = 40;
def F_OK = 0;
def main() -> i64 {
	var filename = "lorem-ipsum.txt\x00";
	var ok_access = wali.SYS_access(Pointer.atContents(filename), F_OK);
	if (ok_access == 0) {
		print("Reading 40 bytes of lorem-ipsum.txt:\n");
		var fd = wali.SYS_open(Pointer.atContents(filename), O_RDONLY, 0);
		var read_result = readfd(i32.view(fd), READ_SIZE);
		print(read_result);
		wali.SYS_close(i32.view(fd));
		return 0;
	}
	print("No access rights to lorem-ipsum.txt:\n");
	return 1;
}

def readfd(fd: i32, length: i32) -> string {
	var str = Array<byte>.new(length);
	wali.SYS_read(i32.view(fd), Pointer.atContents(str), length);
	wali.SYS_close(i32.view(fd));
	return str;
}

def print(str: string) {
	wali.SYS_write(STDOUT, Pointer.atContents(str), str.length);
}
