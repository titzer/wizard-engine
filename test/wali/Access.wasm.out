Reading 40 bytes of lorem-ipsum.txt:
Lorem ipsum dolor sit amet, consectetur 